### PR TITLE
expose metrics scope in workflow and de-dup metrics in replay mode

### PR DESCRIPTION
**Pull request by meiliang86**

### CODE REVIEW

File: src/test/java/com/uber/cadence/workflow/MetricsTest.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    @Override
+    public void execute() {
+      Workflow.getMetricsScope().counter("test-started").inc(1);
+
+      ChildWorkflowOptions options =
+          new ChildWorkflowOptions.Builder().setTaskList(taskList).build();
+      TestChildWorkflow workflow = Workflow.newChildWorkflowStub(TestChildWorkflow.class, options);
+      workflow.executeChild();
+
+      Workflow.getMetricsScope().counter("test-done").inc(1);
```
Comment (mfateev) [inline src/test/java/com/uber/cadence/workflow/MetricsTest.java]: Add a test of the workflow sleeping for some time and a metric measuring the latency of the sleep.

File: src/main/java/com/uber/cadence/internal/metrics/ReplayAwareScope.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+      gauge.update(value);
+    }
+  }
+
+  private static class ReplayAwareTimer implements Timer {
+    Timer timer;
+    ReplayAware context;
+
+    public ReplayAwareTimer(Timer timer, ReplayAware context) {
+      this.timer = Objects.requireNonNull(timer);
```
Comment (mfateev) [inline src/main/java/com/uber/cadence/internal/metrics/ReplayAwareScope.java]: This is not going to work. The problem is that this implementation reuses the timer from the wrapped scope. And the TimerImpl.start calls into the system clock. So during replay the start time will be the time of the replay, not the time the start was called during the first execution. 
Make sure that Workflow.currentTimeMillis() is used everywhere when calculating latencies.

File: src/main/java/com/uber/cadence/internal/replay/DecisionContextImpl.java [hunk 3]
```diff
@@ -40,6 +42,8 @@
 
   private final WorkflowContext workflowContext;
 
+  private Scope metricsScope;
```
[the remark this answers is not in the document]
Comment (meiliang86) [reply src/main/java/com/uber/cadence/internal/replay/DecisionContextImpl.java]: I don't want to leak this pointer before constructor exits.

File: src/main/java/com/uber/cadence/workflow/Workflow.java [hunk 4]
```diff
@@ -708,6 +709,10 @@ public static boolean isReplaying() {
     return WorkflowInternal.isReplaying();
   }
 
+  public static Scope getMetricsScope() {
```
[the remark this answers is not in the document]
Comment (meiliang86) [reply src/main/java/com/uber/cadence/workflow/Workflow.java]: Done.

File: src/test/java/com/uber/cadence/workflow/MetricsTest.java (hunk 1 above)
Comment (meiliang86) [reply src/test/java/com/uber/cadence/workflow/MetricsTest.java]: Done.

File: src/main/java/com/uber/cadence/internal/metrics/ReplayAwareScope.java (hunk 2 above)
Comment (meiliang86) [reply src/main/java/com/uber/cadence/internal/metrics/ReplayAwareScope.java]: Fixed.

File: src/test/java/com/uber/cadence/workflow/MetricsTest.java [hunk 5]
```diff
@@ -83,5 +110,13 @@ public void testWorkflowMetrics() throws InterruptedException {
     verify(reporter, times(1)).reportCounter("test-done", null, 1);
     verify(reporter, times(1)).reportCounter("test-child-started", null, 1);
     verify(reporter, times(1)).reportCounter("test-child-done", null, 1);
+
```
Comment (mfateev) [inline src/test/java/com/uber/cadence/workflow/MetricsTest.java]: Why do you need that sleep above? 
In any case never use Thread.sleep in workflow unit tests. Use env.sleep instead.

Comment (meiliang86) [reply src/test/java/com/uber/cadence/workflow/MetricsTest.java]: This is for the metrics client to flush the cached values out. It's not related to workflow testing.

**Pull request merged**